common/types/string.go

```diff
@@ -221,3 +228,101 @@ func stringStartsWith(s String, pre ref.Val) ref.Val {
 	}
 	return Bool(strings.HasPrefix(string(s), string(preStr)))
 }
+
+func stringFormat(s String, args ref.Val) ref.Val {
+	formatStr := string(s)
+	argList, ok := args.(traits.Lister)
+	if !ok {
+		return MaybeNoSuchOverloadErr(argList)
+	}
+	i := 0
+	var builtStr strings.Builder
+	for i < len(formatStr) {
+		if formatStr[i] == '{' {
+			if i+1 < len(formatStr) && formatStr[i+1] == '{' {
+				err := builtStr.WriteByte('{')
+				if err != nil {
+					return NewErr("error writing format string: %s", err.Error())
+				}
+				i += 2
+				continue
+			}
+			newIndex, val, refErr := parseFormatArg(i, formatStr, argList)
+			if refErr != nil {
+				return refErr
+			}
+			_, err := builtStr.WriteString(val)
+			if err != nil {
+				return NewErr("error writing format string: %s", err.Error())
+			}
+			i = newIndex
+		} else if formatStr[i] == '}' {
+			if i+1 < len(formatStr) && formatStr[i+1] == '}' {
+				err := builtStr.WriteByte('}')
+				if err != nil {
+					return NewErr("error writing format string: %s", err.Error())
+				}
+				i += 2
+				continue
+			} else {
```

Comment:
nit: In above `formatStr[i] == '{'` condition, you don't use `if/else` because you short circuit with `continue` but here you both use `if/else` and a continue to short circuit. Be consistent?

Reply:
If I'm understanding correctly here you're referring to how the `else if formatStr[i] == '}' ` check has a `return NewErr("unmatched '}'")`, but the `if formatStr[i] == '{'` check doesn't have a similar `NewErr` line, right?   For the `}` case there's no valid reason to have a lone `}` in the string, so we know we can immediately return an error.

For the `{` case, if the next `if` isn't true (`if i+1 < len(formatStr) && formatStr[i+1] == '{'`) we're either at the end of the string or the next character isn't also `{`, so we call into `parseFormatArg` since it handles all possible cases (end of string, `{<n>}` substitution, and an unmatched `{`) and returns errors as appropriate. I feel like it's reasonable to have the resulting asymmetry since the formatting grammar is a bit different for `{` and `}` here.

Reply:
I'm suggesting something much simpler. You have two code flows both in for loops, one like `if ... { <do something> continue } <do something else>` and another like `if ... { <do something> continue } else { <do something else }`.  Both are equivalent. Pick a style throughout the code?

Reply:
Oh, I misunderstood, sorry. I'll refactor and mark as resolved.